File: cpp/dynamic-programming/easy/977.有序数组的平方.cpp
/*
 * @lc app=leetcode.cn id=977 lang=cpp
 *
 * [977] 有序数组的平方
 */

// @lc code=start
class Solution
{
public:
    vector<int> sortedSquares(vector<int>& nums)
    {
        int              left   = 0;
        int              right  = nums.size() - 1;
        int              cursor = nums.size() - 1;
        std::vector<int> targets(nums.size(), 0);

        while (left <= right) {
            int powLeft  = pow(nums[left], 2);
            int powRight = pow(nums[right], 2);
            if (powLeft < powRight) {
                targets[cursor] = powRight;
                right--;
            } else {
                targets[cursor] = powLeft;
                left++;
            }
            cursor--;
        }
        return targets;
    }
};
// @lc code=end

File: golang/array/easy/977.有序数组的平方.go
/*
 * @lc app=leetcode.cn id=977 lang=golang
 * @lcpr version=30202
 *
 * [977] 有序数组的平方
 */

// @lc code=start
func sortedSquares(nums []int) []int {
	left, right := 0, len(nums)-1
	res := make([]int, len(nums))
	pos := len(nums) - 1
	for left <= right {
		if nums[left]*nums[left] > nums[right]*nums[right] {
			res[pos] = nums[left] * nums[left]
			left++
		} else {
			res[pos] = nums[right] * nums[right]
			right--
		}
		pos--
	}

	return res
}

// @lc code=end

/*
// @lcpr case=start
// [-4,-1,0,3,10]\n
// @lcpr case=end

// @lcpr case=start
// [-7,-3,2,3,11]\n
// @lcpr case=end

*/

